### PR TITLE
amended `_write_electrodes_tsv` to exclude writing stim channels to electrodes.tsv

### LINKED ISSUE #1022: `write_raw_bids` shoudn't write stim channels to `electrodes.tsv`
I suppose I will label this as a bug, as the current behaviour is inconsistent with the BIDS standard, per discussion in #1006 

`write_raw_bids` shouldn't write stim channels to the `electrodes.tsv` sidecar file, because stim channels don't have associated electrodes.

I can amend the code in `_write_electrodes_tsv`, and submit a PR. I will suggest something like:
```
if ch['kind'] == FIFF.FIFFV_STIM_CH:
    continue
```
at the beginning of this for loop:
https://github.com/mne-tools/mne-bids/blob/54d50f61e01424081793db7ef3f68659319056cc/mne_bids/dig.py#L140-L152

### PR DESCRIPTION
**Pull request by scott-huberty**
closes #1022 

`make tests` and `make pep` passed on my local machine, fingers crossed!

### CODE REVIEW

**Comment by hoechenberger** [comment]
Great, can you add a test and a changelog entry please?

**Review by sappelhoff — APPROVED** [review]
Thanks! +1 to merge once the changelog entry is there

**Comment by scott-huberty** [comment]
Sure! I was going to add a test for this new behaviour using the 'test_egi.mff' in PR #1006 , but you're right it makes more sense to do it here. 

Will add the test and update changelog now.

File: mne_bids/tests/test_dig.py [hunk 1]
```diff
@@ -325,3 +325,38 @@ def test_convert_montage():
...
+    raw = _load_raw()
+    raw.pick_types(eeg=True, stim=True)  # we don't need meg channels
+    bids_root = tmp_path / 'bids1'
+    bids_path = _bids_path.copy().update(root=bids_root, datatype='eeg')
+    os.makedirs(op.join(bids_root, 'sub-01', 'ses-01', 'eeg'), exist_ok=True)
+    write_raw_bids(raw, bids_path)
+
+    # test 1
+    with open(bids_path.directory /
+              'sub-01_ses-01_acq-01_space-CapTrak_electrodes.tsv') as sidecar:
```
Comment (hoechenberger) [inline mne_bids/tests/test_dig.py:342]: Should specify `encoding='utf-8'`

**Comment by scott-huberty** [comment]
I don't see a `latest.inc` file to add my changes to, should I add them `whats_new.rst`? 


I'm not sure if my test is too convoluted, I was trying to test for two cases:

1. for the test file 'sample_audvis_trunc_raw.fif', only eeg channels (with electrodes) should be written to `electrodes.tsv`.
2. The warning that was identified  in #1006 that was getting raised during `read_raw_bids` "There are channels without locations", should no longer be getting raised, so if it does, the test will fail.

File: mne_bids/tests/test_dig.py [hunk 2]
```diff
@@ -325,3 +325,38 @@ def test_convert_montage():
...
+        n_entries = len([line for line in sidecar
+                         if 'name' not in line])  # don't need the header
+        # only eeg chs w/ electrode pos should be written to electrodes.tsv
+        assert n_entries == len(raw.copy().pick_types(eeg=True)
+                                          .info['ch_names'])
+
+    # test 2
+    with pytest.warns(RuntimeWarning) as record:
+        read_raw_bids(bids_path)
+        # list of warnings emitted by read_raw_bids
```
Comment (hoechenberger) [inline mne_bids/tests/test_dig.py:352]: I appreciate your writing this test (looks like a lot of work!), but I don't think we need this one … we just need to ensure we're writing the correct stuff (which the previous test ensures); testing the reading operation again is, IMHO, overkill. I'd suggest to remove `test 2`.
@sappelhoff WDYT?

File: mne_bids/tests/test_dig.py [hunk 3]
```diff
@@ -325,3 +325,38 @@ def test_convert_montage():
     assert pos['coord_frame'] == 'mri'
     assert_almost_equal(pos['ch_pos']['EEG 001'],
                         [-0.0313669, 0.0540269, 0.0949191])
+
+
+def test_electrodes_io(tmp_path):
+    '''test that only channels with associated electrodes were written to
+       electrodes sidecar.'''
```
Comment (hoechenberger) [inline mne_bids/tests/test_dig.py:332]:
```suggestion
    """Ensure only electrodes end up in *_electrodes.json."""
```

File: mne_bids/tests/test_dig.py (hunk 2 above)
Comment (scott-huberty) [reply mne_bids/tests/test_dig.py:352]: Yeah, I was wondering if it was too much ha. I can remove it.

File: mne_bids/tests/test_dig.py [hunk 4]
```diff
@@ -325,3 +325,38 @@ def test_convert_montage():
...
+    raw = _load_raw()
+    raw.pick_types(eeg=True, stim=True)  # we don't need meg channels
+    bids_root = tmp_path / 'bids1'
+    bids_path = _bids_path.copy().update(root=bids_root, datatype='eeg')
+    os.makedirs(op.join(bids_root, 'sub-01', 'ses-01', 'eeg'), exist_ok=True)
```
Comment (hoechenberger) [inline mne_bids/tests/test_dig.py:337]: Use pathlib:

```suggestion
    (bids_root / 'sub-01' / 'ses-01' / 'eeg').mkdir(exist_ok=True)
```

File: mne_bids/tests/test_dig.py [hunk 5]
```diff
@@ -325,3 +325,38 @@ def test_convert_montage():
...
+    write_raw_bids(raw, bids_path)
+
+    # test 1
+    with open(bids_path.directory /
+              'sub-01_ses-01_acq-01_space-CapTrak_electrodes.tsv') as sidecar:
+        n_entries = len([line for line in sidecar
+                         if 'name' not in line])  # don't need the header
+        # only eeg chs w/ electrode pos should be written to electrodes.tsv
+        assert n_entries == len(raw.copy().pick_types(eeg=True)
+                                          .info['ch_names'])
```
Comment (hoechenberger) [inline mne_bids/tests/test_dig.py:347]: I find this indentation a little odd.

How about the very explicit and readable:
```suggestion
        assert n_entries == len(
            raw
            .copy()
            .pick_types(eeg=True)
            .info['ch_names']
        )
```

File: mne_bids/tests/test_dig.py [hunk 6]
```diff
@@ -334,7 +334,7 @@ def test_electrodes_io(tmp_path):
     raw.pick_types(eeg=True, stim=True)  # we don't need meg channels
     bids_root = tmp_path / 'bids1'
     bids_path = _bids_path.copy().update(root=bids_root, datatype='eeg')
-    os.makedirs(op.join(bids_root, 'sub-01', 'ses-01', 'eeg'), exist_ok=True)
+    (bids_root / 'sub-01' / 'ses-01' / 'eeg').mkdir(exist_ok=True)
```
Comment (hoechenberger) [inline mne_bids/tests/test_dig.py:337]: Why do you need to create this directory manually, anyway..?

**Comment by mscheltienne** [comment]
And yes, for the changelog it's in `whats_new.rst`, thanks!

File: mne_bids/tests/test_dig.py (hunk 6 above)
Comment (scott-huberty) [reply mne_bids/tests/test_dig.py:336]: I'm not sure.. I was having a hard time understanding this, but was trying to follow the examples from the other tests in `test.dig.py`

File: mne_bids/tests/test_dig.py (hunk 5 above)
Comment (drammock) [reply mne_bids/tests/test_dig.py]: alternate way that doesn't require copying `raw`: use `raw.get_channel_types()` and count the instances of `'eeg'`

**Review by sappelhoff — COMMENTED** [review]
@scott-huberty please also comment in https://github.com/mne-tools/mne-bids/pull/1028 with your full name, ORCID, and affiliation -- I'd like to add you there.

File: mne_bids/tests/test_dig.py [hunk 7]
```diff
@@ -325,3 +325,25 @@ def test_convert_montage():
...
+              'sub-01_ses-01_acq-01_space-CapTrak_electrodes.tsv') as sidecar:
+        n_entries = len([line for line in sidecar
+                         if 'name' not in line])  # don't need the header
+        # only eeg chs w/ electrode pos should be written to electrodes.tsv
+        assert n_entries == len(
+            raw
+            .copy()
+            .pick_types(eeg=True)
+            .ch_names
+        )
```
Comment (drammock) [inline mne_bids/tests/test_dig.py:349]: seems [my prior comment](https://github.com/mne-tools/mne-bids/pull/1023#discussion_r933557836) got lost: you should not need to do `raw.copy()` here. use `raw.get_channel_types()` and count the `'eeg'`s

File: mne_bids/tests/test_dig.py (hunk 6 above)
Comment (mscheltienne) [reply mne_bids/tests/test_dig.py:336]: The `write_raw_bids` should create the directory tree, could you try without?

File: mne_bids/tests/test_dig.py (hunk 7 above)
Comment (hoechenberger) [reply mne_bids/tests/test_dig.py:349]: Nice trick, @drammock!

Comment (scott-huberty) [reply mne_bids/tests/test_dig.py]: awesome, thanks @drammock!

File: mne_bids/tests/test_dig.py [hunk 8]
```diff
@@ -325,3 +325,24 @@ def test_convert_montage():
...
+
+    with open(bids_path.directory /
+              'sub-01_ses-01_acq-01_space-CapTrak_electrodes.tsv') as sidecar:
+        n_entries = len([line for line in sidecar
+                         if 'name' not in line])  # don't need the header
+        # only eeg chs w/ electrode pos should be written to electrodes.tsv
+        assert n_entries == len(
+            raw
+            .get_channel_types('eeg')
+        )
```
Comment (mscheltienne) [inline mne_bids/tests/test_dig.py:348]:
```suggestion
        assert n_entries == len(raw.get_channel_types('eeg'))
```

Comment (scott-huberty) [reply mne_bids/tests/test_dig.py]: Thanks, @mscheltienne !

File: doc/whats_new.rst [hunk 9]
```diff
@@ -71,6 +71,7 @@ Detailed list of changes
 - :func:`~mne_bids.copyfiles.copyfile_brainvision` can now deal with ``.dat`` file extension, by `Dominik Welke`_ (:gh:`1008`)
 
 - :func:`~mne_bids.print_dir_tree` now correctly expands ``~`` to the user's home directory, by `Richard Höchenberger`_ (:gh:`1013`)
+- :func:`~mne_bids.write_raw_bids` now correctly excludes stim channels when writing to electrodes.tsv, by `Scott Huberty_` (:gh:`1023`)
```
Comment (mscheltienne) [inline doc/whats_new.rst:74]:
```suggestion

- :func:`~mne_bids.write_raw_bids` now correctly excludes stim channels when writing to electrodes.tsv, by `Scott Huberty_` (:gh:`1023`)
```

A blank line separates entries in the changelogs.

File: mne_bids/tests/test_dig.py [hunk 10]
```diff
@@ -325,3 +325,21 @@ def test_convert_montage():
...
+
+
+def test_electrodes_io(tmp_path):
+    """Ensure only electrodes end up in *_electrodes.json."""
+    raw = _load_raw()
+    raw.pick_types(eeg=True, stim=True)  # we don't need meg channels
+    bids_root = tmp_path / 'bids1'
+    bids_path = _bids_path.copy().update(root=bids_root, datatype='eeg')
+    (bids_root / 'sub-01' / 'ses-01' / 'eeg').mkdir(parents=True,
+                                                    exist_ok=True)
```
Comment (mscheltienne) [inline mne_bids/tests/test_dig.py:337]:
```suggestion
```

As written in the comment above, let's try without creating the folder manually, `write_raw_bids` should do it.

**Review by mscheltienne — COMMENTED** [review]
I think those are the last points I see, + the `encoding='utf-8'` to be added as mentioned by Richard.

I hope this is the last change we need to add support for the EGI format!

File: doc/whats_new.rst (hunk 9 above)
Comment (scott-huberty) [reply doc/whats_new.rst:74]: Okay noted it, thanks for fixing that.

File: mne_bids/tests/test_dig.py (hunk 10 above)
Comment (scott-huberty) [reply mne_bids/tests/test_dig.py:337]: Ok! wil commit this and see if it passes the tests. and I will add the UTF-8 encoding, thank you for reminding me about that.

File: mne_bids/tests/test_dig.py (hunk 1 above)
Comment (scott-huberty) [reply mne_bids/tests/test_dig.py]: done, thanks for the tip!

**Comment by hoechenberger** [comment]
Thanks @scott-huberty!

**Comment by scott-huberty** [comment]
> Thanks @scott-huberty!

Thanks everyone for the help!

**Pull request merged**